core/server/master/src/main/java/alluxio/master/file/meta/LockedInodePath.java

```diff
@@ -93,16 +93,41 @@ public synchronized  InodeFile getInodeFile() throws FileDoesNotExistException {
    */
   public synchronized  InodeDirectory getParentInodeDirectory()
       throws InvalidPathException, FileDoesNotExistException {
-    if (mPathComponents.length < 2 || mInodes.size() < (mPathComponents.length - 1)) {
-      // The path is only the root, or the list of inodes is not long enough to contain the parent
+    InodeDirectory inodeDirectory = (InodeDirectory) peekInode(false);
+    if (inodeDirectory == null) {
       throw new FileDoesNotExistException(
           ExceptionMessage.PATH_DOES_NOT_EXIST.getMessage(mUri.getParent()));
     }
-    Inode<?> inode = mInodes.get(mPathComponents.length - 2);
-    if (!inode.isDirectory()) {
-      throw new InvalidPathException(ExceptionMessage.PATH_MUST_HAVE_VALID_PARENT.getMessage(mUri));
+    return inodeDirectory;
+  }
+
+  /**
+   * Peeks an inode on the inode path. The choice of which inode depends on whether the request
+   * comes from a command with recursive flag or not.
+   *
+   * @param isRecursive true if this peekInode is invoked by a command with recursive flag
+   * @return the last existing inode on the inode path if isRecursive is true, the second
+   *         last inode on the inode path if isRecursive is false
+   * @throws InvalidPathException if the parent exists but is not a directory
+   */
+  public synchronized Inode peekInode(boolean isRecursive)
```

Comment:
The call sites can check the recursive option and then decide which method to call based on that. It's a few extra lines of code at the call site, but the meaning of the code is much clearer, and the call site and `peek` methods make sense individually instead of being closely coupled